pkg/operator/controller/canary/service_test.go

```diff
@@ -50,4 +52,62 @@ func Test_desiredCanaryService(t *testing.T) {
 	if !cmp.Equal(service.OwnerReferences, expectedOwnerRefs) {
 		t.Errorf("expected service owner references %#v, but got %#v", expectedOwnerRefs, service.OwnerReferences)
 	}
+
+	expectedAnnotations := map[string]string{
+		"service.beta.openshift.io/serving-cert-secret-name": "canary-serving-cert",
+	}
+	if !cmp.Equal(service.Annotations, expectedAnnotations) {
+		t.Errorf("expected service annotations to be %q, but got %q", expectedAnnotations, service.Annotations)
+	}
+}
+
+func Test_canaryServiceChanged(t *testing.T) {
+	testcases := []struct {
+		description string
+		mutate      func(service *corev1.Service)
+		expected    bool
+	}{
+		{
+			description: "no change",
+			mutate:      func(service *corev1.Service) {},
+			expected:    false,
+		},
+		{
+			description: "changed annotation",
+			mutate: func(service *corev1.Service) {
+				service.Annotations["foo"] = "bar"
+			},
+			expected: true,
+		},
+		{
+			description: "changed ports",
+			mutate: func(service *corev1.Service) {
+				service.Spec.Ports = append(service.Spec.Ports, corev1.ServicePort{
+					Name:       "",
+					Protocol:   corev1.ProtocolTCP,
+					Port:       int32(100),
+					TargetPort: intstr.FromInt(100),
+				})
+			},
+			expected: true,
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.description, func(t *testing.T) {
+			trueVar := true
+			original := desiredCanaryService(metav1.OwnerReference{
+				APIVersion: "apps/v1",
+				Kind:       "daemonset",
+				Name:       "foo",
+				UID:        "bar",
+				Controller: &trueVar,
+			})
+			mutated := original.DeepCopy()
+			tc.mutate(mutated)
+			changed, _ := canaryServiceChanged(original, mutated)
+			if changed != tc.expected {
+				t.Errorf("expected canaryServiceChanged to be %v, got %v", tc.expected, changed)
+			}
```

Comment:
It's useful to check the returned service to make sure that `canaryServiceChanged` both detects that something changed *and* updates the copy it returns.  We do this with other `Test_fooChanged` tests, and it catches a lot of mistakes.  
```suggestion
			if changed, updated := canaryServiceChanged(original, mutated); changed != tc.expected {
				t.Errorf("expected canaryServiceChanged to be %v, got %v", tc.expected, changed)
			} else if changed {
				if changedAgain, _ := canaryServiceChanged(mutated, updated); changedAgain {
					t.Error("canaryServiceChanged does not behave as a fixed-point function")
				}
			}
```